### ISSUE TITLE
Sorting a large number of integers causes huge memory usage and out of memory error

### ISSUE THREAD

**Issue by Maultasche** [Maultasche/NodeLargeSort#6]
Sorting a very large number of integers causes huge memory usage eventually resulting Javascript heap memory exhaustion. I was able to get this to happen by generating and sorting 100 million integers. The memory usage climbed up to around 2GB before the process failed, indicating the that Javascript heap had no more memory available to it.

Now this is silly because the whole point of sorting integers like this is to avoid using large amounts of memory to store them all.  I'm fairly certain that the cause for the program throwing data at the file stream much faster than it can write that data to disk, causing a huge backup as all the data is buffered in memory until it can be written to disk.

I'll have to change the code to pay attention to the flag that indicates the internal buffer is full, pausing all sorting operations until the file stream has had a chance to drain its internal buffer.

**Comment by Maultasche** [comment]
Issue #2 was caused by the numbers being thrown at the file stream faster than the file stream could write to disk, so I'm pretty sure now that it's the cause of this issue as well.

**Comment by Maultasche** [comment]
It turns out there were several issues here. 

There was the issue with data being written to the file stream faster than it could be written to disk. I fixed that so that we pause writing when the stream buffer is full and resume when the stream buffer has been drained.

However, there was also another issue where chunk files were being written to in parallel rather than in series. This lead to the next chunk being sorted and written to the file before the previous chunk had finished writing to the file. As a result, there were numerous integer chunks in memory and numerous file writes happening simultaneously, driving up memory usage significantly.

Finally, the writing of the chunk files in parallel also caused issues with too many files open at a time. After about 1700 open files, the operating system complained about too many files being opened, and the process failed. This only happened for large amounts of chunk files.

Chunk files are now written to one at a time, solving the problem of excessive memory usage for large numbers of chunk files and also solving the issue with too many files being open.
